Soundex: Performance Optimizations

michaelclarkcuadrado opened the pull request:
Hello!

I have optimized the soundex function's cpu performance by a little over 10x.

| Current | This PR |
| --- | --- |
| `0.3875 ns/op` | `0.03767 ns/op` |

At your convenience and discretion, please pull the following changes:

Project:
1.  Created a go.mod file to allow use of `go test` at the project level

Soundex:
1. Removed map and replaced with a `switch case` statement for performance, which avoids allocations and map access latency.
2. Switched all letter variables to bytes, since this function only supports ASCII anyway.
3. Removed all function calls in the inner loop. (`toUpper` was especially slow)
4. Used a stringbuilder so that only a single allocation is made per iteration.
5. I have included the bechmark code I used. This was tested against 500k randomly generated english names. I am not including the test set, so I have commented out the benchmark.

xrash commented:
Hi Michael, thanks for your time and effort contributing to smetrics!

I have a few points to make:

1) Can we let go.mod to be introduced in another PR? That's something I've been postponing for ever at this point and shall take a look at it. Also I'm sure others are using outdated versions of Go and setting the minimum version of the package to 1.21.1 wouldn't be ideal.

2) The current code doesn't check for the validity of the first letter; yours, on the other hand, does, and returns `0000` if it's not valid. I appreciate the attention to this detail, but, in order to avoid changing behavior of code already in use, and because `0000` doesn't look like a valid soundex code after all, I'd rather keep the current behavior. In a future version, if we want to handle this situation differently, an interesting approach could be to simply skip anything outside a-z and consider the first valid letter it finds. For reference, this is what the soundex implementation in postgres' fuzzystrmatch extension does. 

3) Regarding the commented code, I'd rather not include it, even more not so in the test file. I can imagine a "benchmark" directory at the root of the project with stuff like this and the data used in benchmarking could be hosted somewhere else. But I'm not even convinced of the value of having benchmark code in the repo as of now. What do you think?

Thanks again!

michaelclarkcuadrado commented:
I believe I've fixed up those three points:

1. Removed go.mod
2. Removed the first letter check. Just like the existing code, I capitalize it if it is lowercase ascii, but otherwise do no validation.
3. I've removed the commented benchmark.
